### DISCUSSION TITLE
Example Rails 7 project with esbuild, SSR and Tailwind CSS

### DISCUSSION THREAD

**Discussion by ganchdev** [reactjs/react-rails#1308] [Show and tell]
Hi all!

I've been thinking of giving this gem a shot for a new project I'm working on, so I decided to dive in and set it all up myself! I'm pretty happy about the result and how straightforward `react-rails` is for server-side rendered React, so I'm going to share this:

[Example Project Repository](https://github.com/naft-a/react-rails-example)

⬆️ This is the example project, and I've added some notes about what's being used.

Initially, I wanted to try to get it up and running with Vite, but encountered some weird issues with the `vite-ruby` gem, so I went for esbuild (via `jsbundling-rails`) instead, which is just as good (and standard now) for new Rails apps. Anything is better than Webpack in the end, and I'm glad I didn't have to fall back to it this time.

I'm open to questions and suggestions on how to improve it.

**Comment by ahangarha** [comment]
CC @justin808 @Judahmeek

**Comment by justin808** [comment]
@naft-a 
1. did you give https://github.com/shakacode/react_on_rails a try as well?
2. Why do you say:

> ruby gem, so I went for esbuild (via jsbundling-rails) instead, which is just as good (and standard now) for new Rails apps. Anything is better than Webpack in the end 

Shakapacker supports https://github.com/shakacode/shakapacker#esbuild-loader-configuration.

**Reply by ganchdev** [reply]
Hi @justin808 

1. Nope, I did not but may spin up a new project to try that one out! 
2. Mostly because I had bad experience with `webpack` in the past - being too flexible also made it too complicated and I just don’t care enough to really understand it. As for `shakapacker/webpacker`  - I just wanted to have that setup as close as possible to the rails defaults which happens to be `jsbundling-rails` 🤷🏽

**Comment by aniruddhaadak80** [comment]
Hi there! Taking a look at your issue regarding `Example Rails 7 project with esbuild, SSR and Tailwind CSS` in the context of `reactjs/react-rails`...

In the `react-rails` ecosystem, this generally traces back to the asset pipeline (Sprockets vs Webpack/esbuild/Vite) interacting with modern React.
1. **Bundling:** If you're encountering hydration mismatches or module loading issues, verify that your entry points in `app/javascript/components` properly export default/named components as expected by the `react_component` helper.
2. **SSR:** If server-side rendering is enabled, ensure no browser-specific APIs (`window`, `document`) are invoked during the initial render phase in your components. Move those inside `useEffect`.
3. **Caches:** A standard recommendation is running `rake tmp:clear` and clearing out `public/packs` or `public/assets` to be absolutely certain it's not a stale bundle.

Hope this gets your Rails/React setup unblocked and running smoothly!
